### ISSUE TITLE
How to train the model

### ISSUE THREAD

**Issue by LGC163** [Omkaarr1/Medical_Image_Classification_Using_CNN#1]
Hi, your work is very good, congratulations on your work on Medical_Image_Classification_Using_CNN, worth learning from.

I would like to know how your model is trained, do you mention it in your work?

I appreciate your advice.

**Comment by Omkaarr1** [comment]
Below is a high-level overview of how the models (for brain tumor, tuberculosis, and pneumonia classification) are typically trained before they are saved as `.h5` files and subsequently loaded in the `main.py` script. Since the repository directly downloads these pretrained models and does not include a dedicated training script in `main.py`, the training process occurs offline (often in a separate Jupyter notebook or Python script). However, based on common practices for Convolutional Neural Network (CNN) model development and the project structure, you can infer the main steps in detail:

---

## 1. Data Acquisition and Preparation

1. **Data Collection**  
   - Gather image datasets for each classification task:
     - **Brain Tumor Dataset** 
     - **Tuberculosis (TB) Dataset**
     - **Pneumonia Dataset** 

2. **Data Splitting**  
   - Typically split each dataset into **training**, **validation**, and **test** sets (e.g., 70% training, 15% validation, and 15% test).  

3. **Data Augmentation** 
   - Image augmentation techniques (random rotations, flips, zooms, etc.) are often applied to improve model robustness and reduce overfitting.  

4. **Preprocessing**  
   - **Resizing**: Each CNN may require a specific input size (e.g., 224×224 for the brain tumor model, 64×64 for the TB model, 150×150 for the pneumonia model).  
   - **Normalization**: Scaling pixel values to the range [0, 1] or using mean subtraction/standard deviation.  
   - **Color Conversion**: Depending on the model, images may be converted to grayscale or kept in RGB.

---

## 2. Model Architecture

### Convolutional Neural Networks (CNNs)
A typical CNN architecture for medical image classification includes:

1. **Input Layer**  
   - Accepts images of a defined size (e.g., 224×224×3 for color images).

2. **Convolutional Layers**  
   - Extract spatial features by applying convolution filters.  
   - Followed by activation functions such as ReLU to introduce nonlinearity.

3. **Pooling Layers**  
   - Downsample feature maps (e.g., using MaxPooling), reducing spatial dimensions while retaining key features.

4. **Fully Connected (Dense) Layers**  
   - Flatten feature maps into a 1D vector and pass them through one or more Dense layers to learn complex patterns.

5. **Output Layer**  
   - For multi-class classification (e.g., brain tumor types), the final Dense layer often has a Softmax activation to output probabilities for each class.  
   - For binary classification (normal vs. disease), a single neuron with a Sigmoid activation is commonly used.

---

## 3. Compilation and Hyperparameter Tuning

1. **Loss Function**  
   - **Multi-class**: Categorical Crossentropy or Sparse Categorical Crossentropy.  
   - **Binary**: Binary Crossentropy.

2. **Optimizer**  
   - Commonly used optimizers include **Adam**, **SGD**, or **RMSProp**.  
   - Learning rate tuning is critical (e.g., default 1e-3 for Adam).

3. **Metrics**  
   - Typically use **accuracy** for classification tasks.  
   - May also track precision, recall, or F1-score to evaluate performance on imbalanced datasets.

4. **Batch Size and Epochs**  
   - Batch size commonly varies from 16 to 64, depending on GPU/CPU resources.  
   - Number of epochs can range from 10 to 50+; early stopping is often used if validation accuracy stops improving.

5. **Hyperparameter Tuning**  
   - Adjust network depth, number of filters, kernel size, learning rate, and regularization techniques (dropout, L2 regularization) to improve performance.

---

## 4. Model Training

1. **Model Fitting**  
   - Call `model.fit(...)` or a similar method in TensorFlow/Keras with the training and validation sets.  
   - Monitor metrics (accuracy, loss) on both training and validation sets.

2. **Callbacks** (optional but common)  
   - **Early Stopping**: Stop training if validation metrics stop improving.  
   - **Model Checkpoint**: Save the best model weights during training.

3. **Tracking Progress**  
   - Plot training/validation accuracy and loss curves to diagnose overfitting or underfitting.

---

## 5. Evaluation and Validation

1. **Test Set Evaluation**  
   - After training, load the best model weights and evaluate on the **unseen test set** to get final metrics.  
   - Compute accuracy, confusion matrix, precision, recall, and F1-score.

2. **Cross-Validation** (optional)  
   - In some cases, K-fold cross-validation is used to ensure the model generalizes well.

3. **Interpretability**  
   - Grad-CAM or saliency maps may be used to visualize which areas of the image the model relies on for its predictions.

---

## 6. Saving the Model

After achieving satisfactory results, the final model is saved in `.h5` format using:
```python
model.save("pneumonia_model.h5")
```
(or similarly for the other models). These `.h5` files are then uploaded to a shared drive (Google Drive in your case).

---

## 7. Loading and Deployment (As Seen in `main.py`)

In the provided `main.py`, the pre-trained models (`braintumor.h5`, `Tuberculosis_model.h5`, and `pneumonia_model.h5`) are:

1. **Downloaded** if they do not exist locally (using `gdown`).  
2. **Loaded** into memory via `tf.keras.models.load_model(...)`.  
3. **Used** for inference when an uploaded image is passed through a corresponding preprocessing pipeline and then `model.predict(...)`.

This separation of concerns (training offline, then loading for inference) is typical for applications where training takes a long time or requires specialized hardware.

---

## Summary

1. **Data** is curated, split, and preprocessed.  
2. **CNN Architectures** are defined for each condition (brain tumor, TB, pneumonia).  
3. **Hyperparameters** (optimizer, learning rate, epochs, batch size) are chosen and tuned.  
4. **Training** is performed in Keras/TensorFlow, often with callbacks like early stopping.  
5. **Evaluation** occurs on a held-out test set, measuring accuracy and other relevant metrics.  
6. **Saving** the best performing model into `.h5` format.  
7. **Loading** the final model in `main.py` to perform real-time image classification via FastAPI.  

These steps ensure that each of the three pretrained models is robust and accurate before being integrated into the FastAPI application for medical image classification.
